book/src/rfc/rfc-0-generic-parameters.md

```diff
@@ -0,0 +1,359 @@
+    assert_eq(val1, val2);
+}
+```
+
+
+The values for the generic parameters will be passed to the function via the `generics` argument:
+
+```rust
+fn to_bits<B: Backend>(
+    compiler: &mut CircuitWriter<B>,
+    generics: &GenericParameters,
+    vars: &[VarInfo<B::Field, B::Var>],
+    span: Span,
+) -> Result<Option<Var<B::Field, B::Var>>> {
+    ...
+
+    // retrieve the generic values from the `generics` argument
+    let bitlen = generics.get("LEN") as usize;
+
+    ...
+}
+```
+
+Both the return type and returned vars can be checked outside of the builtin functions. The return type can be checked automatically in the same way as the native functions, the types of which are propagated and converged at certain point, at which error will be thrown if the types are not matched.
+
+The return vars can be checked by relying on the types. Each concrete type has a fixed number of vars. With the resolved return type of the builtin function, we can check if the size is matched. Additionally, we can check the values recursively with the type structure, but it might only limited to checking the boolean type which got obvious bound 0 or 1. So automatically checking if the actual return from a builtin is an area to be improved in the future.
+
+
+## monomorphization
+The resolving of the generic values can be done by observing the arguments passed to the function. Then it stores the resolved values in the relevant contexts for the following compiler pipeline to do type checking and circuit synthesizing. We call the process of resolving the generic values and type checking _Monomorphization_.
+
+The current pipeline of compiling noname code is:
+1. Parse the code into AST
+2. Convert the AST into NAST with naming resolution
+3. Convert the NAST into TAST with type metadata collection and type checking
+4. Circuit synthesizing TAST into an constraint system
+
+With generic parameters, the current TAST phase can't handle the type checking anymore, because the generic parameters are unknown. For example, it can't type check the array with symbolic size without resolving the values for the generic parameters.
+
+To extend the type checking to support generic parameters, we can add a MAST phase (Monomorphized AST), right after the TAST phase, to resolve the values of the generic parameters. The circuit synthesizer will rely on the MAST instead of the TAST to compile a circuit.
+
+The new pipeline will be: AST Parser -> NAST -> TAST -> MAST -> Circuit synthesizing
+
+## Implementation
+To support the generic syntax as shown in the examples above, we need to make changes to the AST Parser support generic syntax. Furthermore, because the generic parameters can't be resolved at TAST phase, the some type checkings will be less strict and deferred to MAST phase.
+
+The newly added phase MAST will be responsible for resolving the generic values from the observed arguments. It includes type checking on the monomorphized types that are bypass in the TAST phase. 
```

Comment:
I was reading this to try to understand exactly what we can’t typecheck before the MAST pass (and need to defer to the MAST pass). But it’s not entirely clear to me here. Maybe list exactly the checks that we need to do and can’t do at the TAST pass?

Reply:
added